src/renderers/shared/fiber/ReactFiberHostContext.js

```diff
@@ -35,129 +43,87 @@ module.exports = function<T, P, I, TI, C, CX>(
     getRootHostContext,
   } = config;
 
-  // Context stack is reused across the subtrees.
-  // We use a null sentinel on the fiber stack to separate them.
-  let contextFibers : Array<Fiber | null> = [];
-  let contextValues : Array<CX | null> = [];
-  let contextDepth : number = -1;
-  // Current context for fast access.
-  let currentContextValue : CX | null = null;
-  // Current root instance for fast access.
-  let rootInstance : C | null = null;
-  // A stack of outer root instances if we're in a portal.
-  let portalStack : Array<C | null> = [];
-  let portalDepth : number = -1;
+  let contextStackCursor : StackCursor<?CX> = createCursor((null: ?CX));
+  let contextFiberStackCursor : StackCursor<?Fiber> = createCursor((null: ?Fiber));
+  let rootInstanceStackCursor : StackCursor<?C> = createCursor((null: ?C));
 
   function getRootHostContainer() : C {
-    if (rootInstance == null) {
+    if (rootInstanceStackCursor.current == null) {
```

Comment:
Minor nit: could you read `.current` once? I don't know if compiler is smart enough to know it hasn't changed.